doc/src/fpga_tools/opae.io/opae.io.md

```diff
@@ -0,0 +1,308 @@
+PCIe vendor/device ID, a brief description of the device, and the
+driver to which the device is currently bound.
+
+```opae.io``` provide an option to initialize a PCIe device for use with
+the vfio-pci driver. In order for the device CSRs to be accessed from
+user space, the device must first be bound to the vfio-pci driver. This
+is the job of the ```init``` command option.
+
+`opae.io init [-d PCI_ADDR USER:[GROUP]]`
+
+The ```init``` command unbinds the specified device from its current
+driver and binds it to vfio-pci. This creates a new vfio group under
+/dev/vfio. This group path is then used by the ```libopaevfio.so```
+library to interact with the device.
+
+To release the PCIe device from vfio-pci and return it to use with its
+previous driver, the ```release``` command option is used.
+
+`opae.io release [-d PCI_ADDR]`
+
+The ```release``` command option reverses the actions of the last
+```init``` command, releasing the device from vfio-pci and binding
+it to the driver which was bound at the time the ```init``` command
+was issued.
+
+The ```walk``` command option traverses and displays the Device
+Feature List of the given region.
+
+`opae.io walk [-d PCI_ADDR] [-r REGION] [OFFSET] [-u,--show-uuid]`
+
+The various fields of each Device Feature Header are displayed. The
+`--show-uuid` option additionally displays the GUID for each feature.
+OFFSET can be used to specify the beginning of the DFL in the MMIO
+region.
+
+The ```peek``` command option reads and displays a CSR value.
+
+`opae.io peek [-d PCI_ADDR] [-r REGION] OFFSET`
+
+The ```poke``` command option writes a given value to a CSR.
+
+`opae.io poke [-d PCI_ADDR] [-r REGION] OFFSET VALUE`
+
+```opae.io``` can also execute Python scripts from the command line.
+These Python scripts may contain calls to the device built-in
+functions that are available during an interactive session. Refer
+to the description of interactive mode for details.
+
+`opae.io [-d PCI_ADDR] [-r REGION] myscript.py a b c`
+
+In order to enter the interactive mode of ```opae.io```, simply
+invoke it and optionally pass the desired device address and
+MMIO region options.
+
+`opae.io [-d PCI_ADDR] [-r REGION]`
+
+## INTERACTIVE MODE ##
+
+Upon entering interactive mode, ```opae.io``` begins a Python
+interpreter session and displays the command prompt shown below:
+
+0000:3f:00.0[0]>>
+
+The first portion of the prompt shows the address of the active
+PCIe device, here 0000:3f:00.0. The part in square brackets shows
+the active MMIO region, here [0].
+
+The interpreter waits for a valid Python command, then attempts
+to execute the given command in the usual way. The only
+differences between the traditional Python command intepreter and
+```opae.io``` are that opae.io provides 1) the notion of an active
+PCIe device and MMIO region and 2) several built-in functions and
+objects that allow manipulating the active device and MMIO region.
+
+### BUILT-IN FUNCTIONS ###
+
```

Comment:
It might be nice to reiterate the fact that these built in functions run in the context of the current/active device and region. Maybe something like:
```
The built-in functions below operate in the context of the current or active device and MMIO region that have been opened. Calling these functions without opening both the device and the region will result in an exception being thrown.
```

Reply:
ok